README.md
# Hayaku <sup>[1.5.4](https://github.com/hayaku/hayaku/blob/master/CHANGELOG.md)</sup> [![Build Status][build]][build-link]
[build]: https://travis-ci.org/hayaku/hayaku.png?branch=master
[build-link]: https://travis-ci.org/hayaku/hayaku

Hayaku is a bundle of useful scripts aiming for rapid front-end web development.

The main aim of Hayaku is to create the fastest way to write and maintain CSS code in an editor.

# Table of Contents

1. [Install Hayaku for Sublime Text](#install-hayaku-for-sublime-text)

2. [Features](#features)
    - [Smart CSS Abbreviations](#smart-css-abbreviations)
        - [Fuzzy CSS property abbreviations](#fuzzy-css-property-abbreviations)
        - [Smart CSS values abbreviations](#smart-css-values-abbreviations)
        - [Numeric values in abbreviations](#numeric-values-in-abbreviations)
        - [Color values in abbreviations](#color-values-in-abbreviations) with [RGBA values](#rgba-values)
        - [Importance modifier](#importance-modifier)
        - [Default values](#Default-values)
        - [Clipboard defaults](#clipoard-defaults)
    - [Postexpands](#postexpands)
        - [Simple property postexpands](#simple-property-postexpands)
        - [Postexpands for units](#postexpands-for-units)
        - [Postexpands for colors](#postexpands-for-colors)
        - [Postexpand for importance](#postexpand-for-importance)
    - [Creating new CSS rule blocks](#creating-new-css-rule-blocks)
    - [Inline comments](#inline-comments)
    <br/><br/>
    - [Value cycling <sup>new!</sup>](#value-cycling)
        - [Supported value types](#supported-value-types)
        - [Basics](#basics)
        - [Key bindings](#key-bindings)

3. [Settings and Preferences](#settings-and-preferences)
    - [User dictionaries <sup>new!</sup>](#user-dictionaries)
        - [Syntax of user dictionaries](#syntax-of-user-dictionaries)
        - [User dictionary overrides](#user-dictionary-overrides)
    - [User aliases <sup>new!</sup>](#user-aliases)
    - [Using both dictionary and alias <sup>new!</sup>](#using-both-dictionary-and-alias)
    - [Extra scopes for dictionaries and aliases](#extra-scopes-for-dictionaries-and-aliases)
    - [Autoguessing the code style](#autoguessing-the-code-style)
    - [Single code style](#single-code-style)
    - [Automatic new line after expand](#automatic-new-line-after-expand)
    - [Quotes and URLs](#quotes-and-urls)
    - [Units for unitless values](#units-for-unitless-values)
    - [Prefixes options](#prefixes-options)
        - [The aligning for the prefixes](#the-aligning-for-the-prefixes)
        - [Using only specific prefixes](#using-only-specific-prefixes)
    - [Options for colors](#options-for-colors)
        - [Colors' case](#colors-case)
        - [Shorthand colors](#shorthand-colors)
    <br/><br/>

4. [Using Hayaku with CSS Preprocessors](#using-hayaku-with-css-preprocessors)

5. [License and copyrights](#license-and-copyrights)


# Install Hayaku for [Sublime Text](http://www.sublimetext.com/2)

Right now Hayaku is available only for Sublime Text (even for third version!), but when it would be complete, we would port it to some other editors.

#### Using [Package Control](http://wbond.net/sublime_packages/package_control):

1. Run `Package Control: Install Package` command
2. Search for `Hayaku - tools for writing CSS faster` (`Hayaku` should be enough) and wait for it to be installed
3. Restart Sublime Text (required to make default settings for different syntaxes to work)

#### Or manually, using git:

Clone repository into Packages directory (can be found using `Preferences: Browse Packages` command in Sublime Text)
``` sh
git clone git://github.com/hayaku/hayaku.git
```

And then restart Sublime Text.

## Note on autocomplete

**Important:** Hayaku disables the autocomplete for CSS by default. This was made to remove the ambiguosity and confusion that could happen when you'll see one result in autocomplete and would get something different on pressing `tab`.

You can restore the autocomplete by redefining the `auto_complete_selector` setting in your `User/Preferences.sublime-settings` to either the default value:

``` js
{
    "auto_complete_selector": "source - comment, meta.tag - punctuation.definition.tag.begin"
}
```

Or to anything other you'd like.

However, for CSS scopes this would only enable the autocompletes by `enter`, the `tab` autocomplete would still run the Hayaku when possible.

# Features

## Smart CSS Abbreviations

Hayaku is not your average snippet engine. Most of the CSS snippets to date are static — you need to remember all the abbreviations if you want to use them.

Hayaku offers a better and faster way: you don't need to remember anything, you can just try to write the shortest abbreviation for a thing you want to write in CSS — and Hayaku would try to guess it when you press `tab`.

There are a lot of things Hayaku can do in abbeviations, here are some of them:

### Fuzzy CSS property abbreviations

This is the first basic thing: Hayaku don't have any premade snippets for CSS, it have a dictionary with a lot of CSS properties, so when you write some abrakadabra, it tries to parse it and guess what you meant. For most properties those abbreviations could be rather short, but you're not sticked to them: you can write as much letters for a property as you wish.

So, writing `w`, `wi` or `wid` would give you `width`. And don't forget about the fuzzy part: `wdt` and `wdth` would work too.

Sometimes you would guess that some abbreviations must become other things, but in most cases all the variants have some logic beyound. `b` could be expanded to `background` or `border`, but expanded to `bottom` instead — it's becouse all the “sides” values are abbreviated to just one letter: **l**eft,  **r**ight,  **t**op, so  **b**ottom goes by this path.

However, if you feel that some abbreviation just need to be not that is expands to, feel free to [fill up an issue](https://github.com/hayaku/hayaku/issues/new).

### Smart CSS values abbreviations

Here comes the second basic thing of Hayaku, the awesome one. You can expand abbreviations for the property+value parts, but you don't need to use any delimiters in those abbreviations! That's right — you can just write something like `por` and get `position: relative`!

This works also fuzzy, so to get `position: relative` you could use any number of letters: `pore`, `posrel`, `pstnrltv` etc. Also, if you want, you can still use a delimiter — just add a colon between the property and value and get the same result. So, if you want to stick to Zen style, use it — `pos:r` would work as intended. And `p:r` would work too — while `pr` would expand to `padding-right`, adding delimiter could help by removing ambiguity — padding can't have any values containing `r`, so hayaku falls to `position`.

### Numeric values in abbreviations

Hayaku understands a lot of ways of writing numeric abbreviations.

- You can just write a number after abbreviation to treat it as a value: `w10` would expand to `width: 10px` (see? automatic pixels!).

- Negative numbers supported too: `ml-10` would expand to `margin-left: -10px`.

- If you'd write a dot somewhere in abbreviation, Hayaku would guess what you need `em`s, so `w10.5` would expand to `width: 10.5em`.

- There are some abbreviations for some units, like `percents` for `%`, or `.` for em, so `100p` would expand to `100%` and `10.` to `10em`.

- All other units are supported, `h2pt` would expand to `height:2pt` and so on. Fuzzy guess is there too: if you'd want `vh` you could write just `w10h` and get `width: 10vh`.

### Color values in abbreviations

Actually, you can not only expand strings and numbers, you can expand even colors using abbreviations! You can use different ways to achieve that (as anything in Hayaku), so just look at those examples:

- `c0` → `color: #000`
- `cF` → `color: #FFF` (use uppercase to tell Hayaku it's a color)
- `cFA` → `color: #FAFAFA`
- `c#fa` → `color: #FAFAFA` (no need in uppercase if you use `#`)

And, of course, this works everywhere you would expect colors to work, so `brc0` would expand to `border-right-color: #000;`

#### RGBA values

There is also a way to expand `rgba` values for colors — you can either use rgba's alpha after the dot, either use hexadecimal alpha after the full color, if you'd like. This would look like this:

- `c0.5` → `color: rgba(0,0,0,.5)`
- `cF.2` → `color: rgba(255,255,255,.2)`
- `cABCD` → `color: rgba(170,187,204,0.87)`
- `cABC80` → `color: rgba(170,187,204,0.5)`

You can also write just the dot and get the placeholder on the `alpha` part of the `rgba`:

- `cF00.` → `color: rgba(255,0,0,.[5])`

### Importance modifier

A nice little feature: add `!` after abbreviation and get ` !important` at the end. Not that importance is something you would want to use everyday, but still.

`dn!` would give you `display:none !important;`, yeah.

### Automatic vendor prefixes

If you need some vendor prefixes, Hayaku could provide them!

`bra1.5` would expand to this:

``` css
-webkit-border-radius: 1.5em;
        border-radius: 1.5em;
```

Right now there are no prefixes for values (like gradients etc.) but someday they'd be there.

### Default values

If you'd write something that is only a property (as Hayaku would guess), Hayaku would insert a snippet with some default value already selected for you, so you could start writing your own value to replace it or to press `tab` again to keep it and move forward. So, writing `w` would actually expand to `width: [100%]` (braces mean that this value is selected by default).

### Clipboard defaults

Aside from the normal defaults, Hayaku would try to use your clipboard for getting the value from it as the default value.

Right now it's available for colors and images urls:

- If you'd have color in hexadecimal, rgb(a) or hsl(a) in your clipboard, Hayaku would use it as a default shown value. That would work even is the value is hashless, so if you've copied `808080` from anywhere, then on expanding `c` you would get `color: #[808080]`.

- If you'd have an image url in your clipboard (even relative, Hayaku would look at extension), you'd have it added as default values along inside an `url()`. Also, see [quotes and URLs](#quotes-and-urls) settings on how to adjust the quoting of the inserted url if you want.

#### Configure clipboard defaults

Hayaku offers a setting to set up the behavior of the Clipboard defaults: `hayaku_CSS_clipboard_defaults`. It is an array of the value types that Hayaku could accept as the defaults. So, to disable all the clipboard defaults you could use this setting:

``` js
{
    "hayaku_CSS_clipboard_defaults": [""]
}
```

## Postexpands

“Postexpands” is a nice Hayaku's feature, that allows you to expand only the property at first and then use instant autocomplete for the values of numbers.

That must be said is that postexpand is a bit different from the usual abbreviation expands — it don't have any fuzzy search inside, so only the first letters matter. However, as you'd use it you would see that it is still a powerfull feature.

### Simple property postexpands

The simplest postexpand feature is autocomplete for the string values of different properties.

If you'd expand some property like `po` to `position: |;`, then you could start writing any of it's values and get they expanded right after the cursor. So, writing `a` would give you `position: a|bsolute;`.

### Postexpands for units

Another postexpand feature would allow you to firstly expand the property that can have numeric values, like `width` and then write only the digits and let Hayaku place the corresponding units automatically.

So, when you expand, for example, `w` to `width: |;`, you'd get different options:

- write any iteger like `10` and you'd get `width: 10|px;`
- write any float like `1.5` and you'd get `width: 1.5|em;`
- write an integer and them `e`, so you'd get `width: 10e|m;`
- if the value have any string values, you can also use them: writing `a` would give you `width: a|uto;`

Negative numbers could still be used and if you'd like any other unit, you could just write it down, the autocompleted units won't bother you.

### Postexpands for colors

As you can use shortcuts to colors in abbreviations, you could also write the color values after expanding only the property. The basics are the same: `color: |;` + `F` would give you `color: #F|FF;`, and so on. You can use or don't use the hash symbol.

Another somewhat obscure (but helpful) feature is postexpand for `rgba` colors. This is triggered by writing the comma after decimal value. There is also a shortcut to the alpha value.

- `color: 255,|` would transform to `color: rgba(255,|255,255,1);`
- `color: 255,.|` would transform to `color: rgba(255,255,255,.|5);`

There are a lot of things we could improve there, so stay tuned.

### Postexpand for importance

If you'd like to make some value important, you could just write the first symbols of `!important` keyword and Hayaku would autocomplete it for you.

### Disabling postexpands

If you'd wish to disable postexpands at all for some reason, you could use this setting for this: `"hayaku_CSS_disable_postexpand": true`

## Creating new CSS rule blocks

In Hayaku there is a simple but powerful feature: when you wrote a selector, you could just press `CMD+Enter` to get a block for writing CSS here.

## Inline comments

Another little helper: write `//` in CSS to have it expanded to `/* | */` (where the pipe is a caret placement).

If you'd wish to disable inline comments, you could use this setting: `"hayaku_CSS_disable_inline_comment": true`

*This feature is in development, we plan on adding a lot of things to make commenting fun.*

## Value cycling

Another big feature Hayaku provides — a powerful implementation of cycling values. You can use it to increment or decrement any numeric values (both in CSS and anywhere else) and to cycle through all the possible values for any given CSS property.

### Supported value types

- **CSS values** — both numeric, like `10px`, and string values, like in `position: fixed` are supported.
- **Dates in ISO format**.
- **Versions by semver**.
- **Just any numbers** — even inside word-like strings like `$foo4`.

### Basics

Unlike other similar implementations, Hayaku's cycling is much more powerful and polished:

- When nothing is selected, Hayaku finds the closest to the caret value in the line and cycles through it.

- When only part of a number is selected, Hayaku understands the context and cycles the selected digit, but handling the whole number, so you could easily increment/decrement numbers by 100, 1000 or 0.001 etc.

- Hayaku tries its best to save the position of the cursor or selection, by adjusting it after the replacement. This way you would always be at the same place as before no matter what.

- Hayaku perfectly handles multiple carets, cycling through each of them.


CHANGELOG.md
# Changelog for Hayaku

## 1.5.4 <sup>2019.06.24</sup>

- Added basic support for CSS embedded inside JS ([#313][]).

[#313]: https://github.com/hayaku/hayaku/pull/313

## 1.5.3 <sup>2016.03.04</sup>

- Added basic support for PostCSS.
- Fixed a case when the `CSS_colors_case` options is not found.

## 1.5.2 <sup>2014.08.06</sup>

- Fixed missing `none` value for border-… shorthands ([#291][]).

[#291]: https://github.com/hayaku/hayaku/pull/291

## 1.5.1 <sup>2014.07.31</sup>

- Fixed `CSS_syntax_quote_symbol` option that didn't work when `CSS_syntax_url_quotes` was set to false.

## 1.5.0 <sup>2014.07.07</sup>

- Added support for any hayaku abbreviations as alias values.
- Added support for any unknown values after a colon in abbreviation.
- Added support for unknown units after numbers.
- Added `plain-text` template output, alongside the `full`.
- Updated and fixed dictionary with some transform and animation stuff.
- Fixed incorrect letter case dependance for values.
- Fixed lost transform alias.
- Fixed troubles with spaces in values.
- Fixed unavailable user preprocessor dicts in cycling.
- Fixed the handling of full properties/values.
- Fixed the colon at the end of abbreviation.
- Fixed absent percents for border-radius.
- Fixed quote symbol option to work in more cases.

## 1.4.2 <sup>2014.06.12</sup>

- Fixed regression for inner tabstops, like in transforms.

## 1.4.1 <sup>2014.06.11</sup>

- Fixed ST2 utf problem.

## 1.4.0 <sup>2014.06.09</sup>

- **New feature:** value cycling.
- **New feature:** user-defined dictionaries.
- **New feature:** user-defined aliases.
- Refactored code a bit, added basic tests.
- Now using commented for abbreviations links from dict in postexpands.
- Updated flex properties and some other minor ones.
- Allowed usage of `+` (although, not in the end), `$` and `@` in abbreviations.
- Fixed rare postexpand conflicts.
- Fixed always positive `hayaku_CSS_prefixes_align` setting.

## 1.3.6 <sup>2014.02.01</sup>

- Better postexpand for colors, using color length option.
- Mentioned disabling autocomplete in README.
- Slightly better disabling of autocomplete.
- Allowed using `enter` for autocomplete.

## 1.3.5 <sup>2013.10.27</sup>

- Fixed incorrect indent in Sass-like syntax for new blocks ([#198][])

[#198]: https://github.com/hayaku/hayaku/issues/198

## 1.3.4 <sup>2013.10.26</sup>

- Fixed support of `.less` files ([#227][])
- Fixed the style of colors inserted from clipboard ([#212][])

[#227]: https://github.com/hayaku/hayaku/issues/227
[#212]: https://github.com/hayaku/hayaku/issues/212

## 1.3.3 <sup>2013.03.02</sup>

- **New setting:** `hayaku_CSS_syntax_quote_symbol` for used quote symbol ([#71][])

- **New setting:** `hayaku_CSS_syntax_url_quotes` for wrapping clipboarded links in urls with quotes ([#208][])

- Added support for `user-select` property ([#207][])

- Fixing values in abbreviations for `opacity` property ([#209][])

- Don't add units if the value could be unitless (like `line-height`), also `hayaku_CSS_units_for_unitless_numbers` setting for this ([#153][])

[#71]: https://github.com/hayaku/hayaku/issues/71
[#207]: https://github.com/hayaku/hayaku/issues/207
[#208]: https://github.com/hayaku/hayaku/issues/208
[#209]: https://github.com/hayaku/hayaku/issues/209
[#153]: https://github.com/hayaku/hayaku/issues/153

## 1.3.2 <sup>2013.02.27</sup>

- Fix the importing method for ST3 again, now should work from the `.sublime-package`.

## 1.3.1 <sup>2013.02.08</sup>

- Fix the importing method, now the plugin would work from PC in ST3 ([#205][])

[#205]: https://github.com/hayaku/hayaku/issues/205

## 1.3.0 <sup>2013.02.07</sup>

- **Support for [Sublime Text 3](http://www.sublimetext.com/3) ([#201][])**

- Fixed inline comment setting for OS X and Linux ([#200][], thanks to @freshmango)

- Disable inline commenting in functions and quotes ([#203][])

[#200]: https://github.com/hayaku/hayaku/issues/200
[#201]: https://github.com/hayaku/hayaku/issues/201
[#203]: https://github.com/hayaku/hayaku/issues/203

## 1.2.1 <sup>2012.12.23</sup>

- Hotfixing automatic new line after expand's bug ([#190][])

[#190]: https://github.com/hayaku/hayaku/issues/190

## 1.2.0 <sup>2012.12.23</sup>

- **New feature:** [basic clipboard defaults](https://github.com/hayaku/hayaku/#clipboard-defaults) (for colors and urls) ([#180][])
- **New setting:** optional [automatic new line after expand](https://github.com/hayaku/hayaku/#automatic-new-line-after-expand) (not by default) ([#123][])
- Better handling of multiple carets in snippets ([#188][])
- Fixed an issue with color postexpands and their default values ([#189][])
- Restructured the repo, so it would be better updatable and maintainable.

[#123]: https://github.com/hayaku/hayaku/issues/123
[#180]: https://github.com/hayaku/hayaku/issues/180
[#188]: https://github.com/hayaku/hayaku/issues/188
[#189]: https://github.com/hayaku/hayaku/issues/189

## 1.1.1 <sup>2012.12.16</sup>

- Fixed bug with more than 99 completion parts in a snippet (`display: inline` affected) ([#182][])
- Fixed bug with wrong position of color postexpand in prefixed clusters ([#183][])
- Better handling for values that can be parts of other values in postexpands ([#184][])
- Overall refactoring of the postexpands, not completed, but already fixed some minor issues and the code is almost ready for moving the postexpands to the dictionaries.

[#182]: https://github.com/hayaku/hayaku/issues/182
[#183]: https://github.com/hayaku/hayaku/issues/183
[#184]: https://github.com/hayaku/hayaku/issues/184

## 1.1.0 <sup>2012.12.10</sup>

- **Changed default setting**: now when you use the block expand it expands to the more common code style.
- **New feature:** added [importance to the postexpand](https://github.com/hayaku/hayaku/#postexpand-for-importance) ([#156][])
- **New setting:** disabling the [inline comment](https://github.com/hayaku/hayaku/#inline-comments) shortcut for CSS ([#169][])

hayaku_probe.py
# -*- coding: utf-8 -*-
# (c) 2012 Sergey Mezentsev
import os
import re
from itertools import product, chain

def import_dir(name, fromlist=()):
    PACKAGE_EXT = '.sublime-package'
    dirname = os.path.basename(os.path.dirname(os.path.realpath(__file__)))
    if dirname.endswith(PACKAGE_EXT):
        dirname = dirname[:-len(PACKAGE_EXT)]
    return __import__('{0}.{1}'.format(dirname, name), fromlist=fromlist)


try:
    imp = import_dir('hayaku_dict_driver', ('css_defaults', 'get_css_dict', 'get_flat_css', 'css_flat_list'))
    css_defaults = imp.css_defaults
    get_css_dict = imp.get_css_dict
    get_flat_css = imp.get_flat_css
    css_flat_list = imp.css_flat_list
except ImportError:
    from hayaku_dict_driver import css_defaults, get_css_dict, get_flat_css, css_flat_list

# TODO: Move this to dicts etc.
PRIORITY_PROPERTIES = [ 'display', 'color', 'margin', 'position', 'padding', 'width', 'background', 'zoom', 'height', 'top', 'vertical-align', 'overflow', 'left', 'margin-right', 'float', 'margin-left', 'cursor', 'text-decoration', 'font-size', 'margin-top', 'border', 'background-position', 'font', 'margin-bottom', 'padding-left', 'right', 'padding-right', 'line-height', 'white-space', 'text-align', 'border-color', 'padding-top', 'z-index', 'border-bottom', 'visibility', 'border-radius', 'padding-bottom', 'font-weight', 'clear', 'max-width', 'border-top', 'border-width', 'content', 'bottom', 'background-color', 'opacity', 'background-image', 'box-shadow', 'border-collapse', 'text-overflow', 'filter', 'border-right', 'text-indent', 'clip', 'min-width', 'min-height', 'border-left', 'max-height', 'border-right-color', 'border-top-color', 'transition', 'resize', 'overflow-x', 'list-style', 'word-wrap', 'border-left-color', 'word-spacing', 'background-repeat', 'user-select', 'border-bottom-color', 'box-sizing', 'border-top-left-radius', 'font-family', 'border-bottom-width', 'outline', 'border-bottom-right-radius', 'border-right-width', 'border-top-width', 'font-style', 'text-transform', 'border-bottom-left-radius', 'border-left-width', 'border-spacing', 'border-style', 'border-top-right-radius', 'text-shadow', 'border-image', 'overflow-y', 'table-layout', 'background-size', 'behavior', 'body', 'name', 'letter-spacing', 'background-clip', 'pointer-events', 'transform', 'counter-reset', ]

def get_all_properties(css_dict=None):
    if css_dict is None:
        css_dict = get_css_dict()[0]

    all_properties = list(css_dict)

    # раширить парами "свойство значение" (например "position absolute")
    for prop_name in all_properties:
        property_values = css_flat_list(prop_name, css_dict)
        extends_sieve = (i for i in property_values if not i[1].startswith('<'))
        unit_sieve = (i for i in extends_sieve if not i[1].startswith('.'))
        all_properties.extend('{0} {1}'.format(prop_name, v[1]) for v in unit_sieve)
    return all_properties


def score(a, b):
    """Оценочная функция"""
    s = 0

    # увеличивает вес свойству со значением (они разделены пробелом)
    if a and ' ' == a[-1]:
        s += 3.0

    # уменьшить, если буква находится не на грницах слова
    if '-' in a[1:-1] or '-' in b[1:-1]:
        s += -2.0

    # уменьшить, если буква находится не на грницах слова
    if ' ' in a[1:-1] or ' ' in b[1:-1]:
        s += -0.5

    # если буква в начале слова после -
    if a and a[-1] == '-':
        s += 1.05

    # если буквы подряд
    if len(a) == 1:
        s += 1.0

        # последовательные буквы
        if len(b) == 1:
            s += 1.0

    return s

def string_score(arr):
    """Получает оценку разбиения"""
    # s = sum(score(arr[i-1], arr[i]) for i in range(1, len(arr)))
    # if s >0 :
    #     print arr, s
    return sum(score(arr[i-1], arr[i]) for i in range(1, len(arr)))

def tree(css_property, abbr):
    # функция генерирует деревья (разбиения) из строки
    # (abvbc, abc) -> [[a, bvb ,c], [avb, b, c]]
    # print '\n', css_property
    if len(css_property) < len(abbr):
        return set([])
    trees = [[css_property[0], css_property[1:],],]
    for level in range(1, len(abbr)):
        # print level, trees
        for tr in trees:
            if level == 1 and len(trees) == 1:
                trees = []
            # находит индексы букв
            indexes = []
            i = -1
            try:
                while True:
                    i = tr[-1].lower().index(abbr[level], i+1)
                    indexes.append(i)
            except ValueError:
                pass
            # print 'indexes len', len(indexes)
            for ind in indexes:
                if level == 1:
                    car = tr[:-1]
                    cdr = tr[-1]
                    first = cdr[:ind]
                    second = cdr[ind:]
                    add = []
                    add.append(car[-1] + first)
                    add.append(second)
                    # print '\t', car, '|', cdr,'|', first,'|', second, '-', add, level, '=', tr
                    trees.append(add)
                else:
                    car = tr[:-1]
                    cdr = tr[-1]
                    first = cdr[:ind]
                    second = cdr[ind:]
                    add = car
                    add.append(first)
                    add.append(second)
                    # print '\t', car, '|', cdr,'|', first,'|', second, '-', add, level, '=', tr
                    # print repr(first)
                    trees.append(add)
                # break
            trees_i = set([tuple(t) for t in trees if len(t) == level+1])
            trees = [list(t) for t in trees_i]
            # print 'trees_i', trees_i
            # break
            # print
        # break

    # удалить разбиения с двумя "-" в шилде
    ret = set([tuple(t) for t in trees])
    filtered = []
    for s in ret: # каждое элемент в сете
        for t in s: # каждый шилд в элементе
            # print '\t', t
            if t.count('-') > 1:
                break
        else:
            filtered.append(s)
    # print set([tuple(t) for t in trees])
    # print filtered
    return filtered


def prop_value(s1, val, all_properties):
    """Генератор возвращает свойства и значения разделённые пробелом
    Из всех свойств выбирает только с совпадающим порядком букв"""
    for pv in all_properties:
        if ' ' not in pv.strip():
            continue
        prop, value = pv.split()
        if sub_string(value.lower(), val.lower()):
            if sub_string(prop.lower(), s1.lower()):
                yield '{0} {1}'.format(prop, value).strip()

def sub_string(string, sub):
    """Функция проверяет, следуют ли буквы в нужном порядке в слове"""
    index = 0
    string = string.lower()
    for c in sub:
        try:
            index += string[index:].index(c)+1
        except ValueError:
            return False
    else:
        return True

def segmentation(abbr):
    """Разбивает абрревиатуру на элементы"""

    # Части аббревиатуры
    parts = {
        'abbr': abbr # todo: выкинуть, используется только в тестах
    }

    # Проверка на important свойство
    if '!' == abbr[-1]:
        abbr = abbr[:-1]
        parts['important'] = True
    else:
        parts['important'] = False

    # TODO: вынести regex в compile
    # todo: начать тестировать regex
    m = re.search(r'^([a-z\$\@\+]?[a-z-]*[a-z]).*$', abbr)
    property_ = m if m is None else m.group(1)
    if property_ is None:
        # Аббревиатура не найдена
        return parts
    # del m

    parts['property-value'] = property_

    # удалить из аббревиатуры property
    abbr = abbr[len(property_):]

    if abbr:
        parts['property-name'] = property_
        del parts['property-value']

    # убрать zen-style разделитель
    if abbr and ':' == abbr[0]:
        abbr = abbr[1:]

        if len(abbr) == 0:
            parts['keyword-value'] = ''

    if not abbr:
        return parts

    parts.update(value_parser(abbr))

    if 'value' in parts:
        assert parts['value'] is None
        del parts['value']
    elif ('type-value' not in parts and 'type-name' not in parts):
        parts['keyword-value'] = abbr

    # TODO: сохранять принимаемые значения, например parts['allow'] = ['<color_values>']
    return parts

def value_parser(abbr):
    # todo: поддержка аббревиатур "w-.e" то есть "width -|em"
    parts = {}

    # Checking the color
    # Better to replace with regex to simplify it
    dot_index = 0
    if '.' in abbr:
        dot_index = abbr.index('.')
    if abbr[0] == '#':
        parts['color'] = (abbr[1:dot_index or 99])
        if dot_index:
            parts['color_alpha'] = (abbr[dot_index:])
        parts['value'] = None
    try:
        if all((c.isupper() or c.isdigit() or c == '.') for c in abbr) and 0 <= int(abbr[:dot_index or 99], 16) <= 0xFFFFFF:
            parts['color'] = abbr[:dot_index or 99]
            if dot_index:
                parts['color_alpha'] = (abbr[dot_index:])
            parts['value'] = None
    except ValueError:
        pass

    # Проверка на цифровое значение
    val = None

    numbers = re.sub("[a-z%]+$", "", abbr)
    try:
        val = float(numbers)
        val = int(numbers)
    except ValueError:
        pass

    if val is not None:
        parts['type-value'] = val
        if abbr != numbers:
            parts['type-name'] = abbr[len(numbers):]

    return parts

def extract(hayaku):
    if isinstance(hayaku, dict):
        s1 = hayaku.get('abbr')
        css_dict = hayaku.get('options').get('dict')
        css_aliases = hayaku.get('options').get('aliases')
    else:
        s1 = hayaku
        css_dict, css_aliases = get_css_dict()

    """В зависимости от найденных компонент в аббревиатуре применяет функцию extract"""
    prop_iter = []
    parts = segmentation(s1)
    abbr_value = False
    if 'property-name' in parts:
        if parts['important']:
            s1 = s1[:-1]
        if s1[-1] != ':' and s1 != parts['property-name']:
            abbr_value = True

    if 'color' in parts:
        prop_iter.extend(prop for prop, val in get_flat_css(css_dict) if val == '<color_values>')

    if isinstance(parts.get('type-value'), int):
        prop_iter.extend(prop for prop, val in get_flat_css(css_dict) if val == '<integer>')

    if isinstance(parts.get('type-value'), float):
        # TODO: добавить deg, grad, time
        prop_iter.extend(prop for prop, val in get_flat_css(css_dict) if val in ('<length>', '<number>', 'percentage'))

    # TODO: проверить, всегда ли эта переменная нужна для следующих условий
    all_properties = get_all_properties(css_dict)

    if 'keyword-value' in parts and not parts['keyword-value']:
        prop_iter.extend(all_properties)

    if 'keyword-value' in parts:
        prop_iter.extend(prop_value(parts['property-name'], parts['keyword-value'], all_properties))
    elif 'color' not in parts or 'type-value' in parts:
        prop_iter.extend(all_properties)

    if not parts or not (parts.get('property-name', '') or parts.get('property-value', '')):
        return

    abbr = ' '.join([
        parts.get('property-name', '') or parts.get('property-value', ''),
        parts.get('keyword-value', ''),
    ])

    abbr = abbr.strip()

    if not css_aliases.get(s1):
        abbr = css_aliases.get(abbr, abbr)
        if abbr[-1] == ':':
            abbr = abbr[:-1]

    starts_properties = []
    # todo: переделать механизм PAIRS
    # надо вынести константы в css-dict
    # по две буквы (bd, bg, ba)
    pair = None
    for alias in css_aliases:
        if (alias.endswith('...')) and abbr.startswith(alias[:-3]):
            pair = css_aliases.get(alias)
            break

    if pair is not None:
        starts_properties = [prop for prop in prop_iter if prop.startswith(pair) and sub_string(prop, abbr)]
    if not starts_properties:
        starts_properties = [prop for prop in prop_iter if prop[0] == abbr[0] and sub_string(prop, abbr)]

    if 'type-value' in parts or ('keyword-value' in parts and parts['keyword-value'] == ''):
        starts_properties = [i for i in starts_properties if ' ' not in i]

    property_ = hayaku_extract(abbr, starts_properties, PRIORITY_PROPERTIES, string_score)
    property_, value = property_.split(' ') if ' ' in property_ else (property_, None)
    # print property_, value
    if not property_:
        return {}

    parts['property-name'] = property_

    if value is not None:
        parts['keyword-value'] = value

    # Проверка соответствия свойства и значения

    allow_values = [val for prop, val in get_flat_css(css_dict) if prop == parts['property-name']]

    if 'color' in parts and '<color_values>' not in allow_values:
        del parts['color']
    if 'type-value' in parts and not any((t in allow_values) for t in ['<integer>', 'percentage', '<length>', '<number>', '<alphavalue>']):
        del parts['type-value']
    if 'keyword-value' in parts and parts['keyword-value'] not in allow_values:
        del parts['keyword-value']

    if all([
            'keyword-value' not in parts,
            'type-value' not in parts,
            'color' not in parts,
        ]) and abbr_value:
        return {}

    # Добавить значение по-умолчанию
    if parts['property-name'] in css_dict:
        default_value = css_defaults(parts['property-name'], css_dict)
        if default_value is not None:
            parts['default-value'] = default_value
        obj = css_dict[parts['property-name']]
        if 'prefixes' in obj:
            parts['prefixes'] = obj['prefixes']
            if 'no-unprefixed-property' in obj:
                parts['no-unprefixed-property'] = obj['no-unprefixed-property']

    if parts['abbr'] == parts.get('property-value'):
        del parts['property-value']

    return parts

def hayaku_extract(abbr, filtered, priority=None, score_func=None):
    # выбирает только те правила куда входят все буквы в нужном порядке

    #  все возможные разбиения
    trees_filtered = []
    for property_ in filtered:
        trees_filtered.extend(tree(property_, abbr))

    # оценки к разбиениям
    if score_func is not None:
        scores = [(score_func(i), i) for i in trees_filtered]

        # выбрать с максимальной оценкой
        if scores:
            max_score = max(s[0] for s in scores)
            filtered_scores = (i for s, i in scores if s == max_score)
            filtered = [''.join(t) for t in filtered_scores]
            if len(filtered) == 1:
                return ''.join(filtered[0])

    # выбрать более приоритетные
    if len(filtered) == 1:
        return filtered[0]
    elif len(filtered) > 1 and priority is not None:
        # выбирает по приоритету
        prior = []
        for f in filtered:
            p = f.split(' ')[0] if ' ' in f else f
            try:
                prior.append((priority.index(p), f))
            except ValueError:
                prior.append((len(priority)+1, f))
        prior.sort()
        try:
            return prior[0][1]
        except IndexError:
            return ''
    else:
        return ''
